### DISCUSSION TITLE
How to improve loading account data: balances, names, and other assets

### DISCUSSION THREAD

**Discussion by 314159265359879** [blockstack/stacks-wallet-web#5886] [Ideas]
This is the core problem I see
I see the wallet loading assets and balances over and over again all the time even when there have been no changes since the last time I looked at the account.
It seems inefficient regardless of if there is a power user behind Leather or not. I think overall the wallet would benefit from more data persistence and loading _based on changes_ rather than starting over and _fetching everything_: all Stacks NFT's and Ordinals etc. again.

Some of this is happening as you switch between accounts, on the select account screen. But it also happens when you remain in the same account. I would love to see a robust and lasting solution for this. Originally I wrote this up as a solution but agree that a more general discussion on how to handle this is a better. 

>Original post
>I think one of the biggest wins would be to 
A. load account data in the background for all accounts in a wallet (very slowly to stay well below the limit and to be able to prioritize getting the latest data in an account that is being switched to) and 
B. making less calls by only getting changes since the last update, rather then everything, which seems to be what is happening now. Perhaps timestamps in blocks can help with that ? https://github.com/stacks-network/stacks-core/issues/4541

**Comment by markmhendrickson** [comment]
@kyranjamie @alter-eggo I'd be curious to get your immediate thoughts on the above proposal. Should we turn this into a discussion rather than an issue? It may affect how we structure mobile in particular.

**Comment by kyranjamie** [comment]
This issue describes a solution, rather than a problem. What problem are we trying to solve here? Account data loading faster?

**Comment by markmhendrickson** [comment]
Yea, I understand the problem for power users is that they have to wait a long time when they try certain operations that require the loading of lots of data (specifically the accounts picker). So @314159265359879 is thinking perhaps we can resolve that wait by preloading the data before they make those operations.

**Comment by kyranjamie** [comment]
While I'm sure there's more we can do to improve the efficiency of our requests, I'm not sold on this being the best solution. 

Requesting additional data during idle time is a legitimate technique, for example [it's used to load JS bundles to improve app load performance in Angular](https://angular.dev/guide/ngmodules/lazy-loading#module-based-application). 

But, unlike JS files, the data discussed here goes stale. That means that, in many cases, we'll load data unnecessarily and by the time the user wants to see it, it'll be stale and need refetching again. Especially on mobile, this is expensive as you're asking the phone to perform battery-draining actions _just in case_ advance users _might_ want to see it.

We already optimise by only loading what's in view, and this goes against that principal. I don't want to go back to requesting 300 account balances at one time. Further, identifying when the app is idle isn't necessarily easy. Maybe we detect the app's idle, but then the user does something, and their experience is degraded because we have so many requests in flight.

The balance doesn't have to have loaded to select the account, so account data loading isn't blocking. If account balances are slow to load, let's cache longer, and increase the stale time as it's probably okay on most occasions.

I welcome the idea, though recommend we begin by identifying the core problem, rather than suggesting a solution up front.

**Comment by 314159265359879** [comment]
I think this is the core problem I see
I see the wallet loading assets and balances over and over again all the time even when there have been no changes since the last time I looked at the account. 
It seems inefficient regardless of if there is a power user behind Leather or not. I think overall the wallet would benefit from more data persistence and loading _changes_ rather than starting over and fetching all Stacks NFT's and Ordinals again, which it does often.

**Comment by markmhendrickson** [comment]
Let's convert this to a discussion? @pete-watters raised a similar question re: querying across accounts recently in Slack

![Image](https://github.com/user-attachments/assets/d3acd849-a625-4daa-8d18-e718dbd6b52b)

**Reply by 314159265359879** [reply]
Good idea, I have converted it to a discussion. And renamed the title to reflect a more open discussion about possible solutions rather than discussing a singular one as Kyran suggested.

**Comment by pete-watters** [comment]
Thanks for moving this @314159265359879 . I opened another discussion about the broader concept of data provision [here](https://github.com/orgs/leather-io/discussions/385).

Once we improve on that we can apply the best solution across all platforms and will keep this discussion in mind when doing that.
